docs/source/cuda/examples.rst

```diff
@@ -3,6 +3,479 @@
+
+.. code-block:: python
+
+   f[1, 3](a, b, c, len(a))
+   print(c.copy_to_host())
+
+This also prints
+
+.. code-block:: none
+
+   [2 4 6]
+
+.. _cuda-laplace:
+
+1D Heat Equation
+=====================
+This example solves Laplace's equation in one dimension for a certain set of initial
+conditions and boundary conditions. A full discussion of Laplace's equation is out of
+scope for this documentation, but it will suffice to say that it describes how heat
+propagates through an object over time. It works by discretizing the problem in two ways:
+
+1. The object is separated into small "pieces" that each have an individual temperature
+2. Time is separated into small "intervals" and the universe advances through them one by one
+
+Then, the following assumption is applied: The temperature of a piece after some interval 
+has passed is some weighted average of the temperature of the pieces that are directly 
+touching it. Intuitively, this means if all the small pieces of the object are very hot 
+and a single piece in the middle is very cold, as time goes on, the hot pieces will cause 
+the cold one to heat up and the cold piece will cause the surrounding hot pieces to cool 
+slightly. Simply put, the heat spreads out around the object.
+
+We can simulate this situation using a Kernel generated from Numba. Let's start simple by assuming
+we have a one dimensional object which we'll represent with an array of values. The position 
+of the element in the array is like the position of a small piece of the object, and the value 
+of the element is like the temperature. 
+
+.. code-block:: python
+
+   from numba import cuda
+   import numpy as np
+
+
+Some initial setup here. Let's make one point in the center of the object very hot.
+
+.. code-block:: python
+
+   size = 100
+   data = np.zeros(size)
+
+   # make one piece in the very middle very hot
+   data[int(size/2)] = 10000
+   data_gpu = cuda.to_device(data)
+
+
+In our kernel each thread will be responsible for managing the temperature update for a single element
+in a loop over the desired number of timesteps. The kernel is below:
+
+.. code-block:: python
+
+   @cuda.jit
+   def solve_heat_equation(data, size, timesteps, k):
+      i = cuda.grid(1)
+
+      for step in range(timesteps):
+         # get the current temperature associated with this segment
+         curr_temp = data[i]
+         
+         # apply formula from finite difference equation
+         if i == 0:
+               # Left wall is held at T = 0
+               next_temp = curr_temp + k * (data[i + 1] - (2 * curr_temp))
+         elif i == size - 1:
+               # Right wall is held at T = 0
+               next_temp = curr_temp + k * (data[i - 1] - (2 * curr_temp))
+         else:
+               next_temp = curr_temp + k * (data[i - 1] - (2 * curr_temp) + data[i + 1])
+         data[i] = next_temp
+         
+         # wait for every thread to write before moving on
+         cuda.syncthreads()
```

Comment:
How does syncing at the block level have any meaningful effect here?

It feels like this example needs two copies of the domain along with a cooperative grid sync, or one kernel launch per timestep in order to ensure that all inputs are taken from the previous timestep and not some values coming from the current timestep at the boundaries between blocks.

Am I misunderstanding the operation of the example?

Reply:
Yeah, this example was one that I remembered from academia. I think I derived this from some memory of a pure c++ example which used a grid wide barrier. This will be updated.